File: test/integration/python/test_mute_layers.py
# -*- coding: utf-8 -*-

from pxr import Usd, Sdf, Tf
from usd_notice_framework import Broker, UnfNotice

import pytest


@pytest.mark.parametrize("notice_type, excepted", [
    ("StageNotice", 12),
    ("StageContentsChanged", 4),
    ("ObjectsChanged", 4),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 4),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_mute_layers(notice_type, excepted, stage_with_layers):
    """Mute several layers.
    """
    stage = stage_with_layers
    Broker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(UnfNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    layers = stage.GetRootLayer().subLayerPaths

    # Keep ref pointer to the layer we try to mute and unmute to
    # prevent it for being destroyed when it is muted.
    _layer = Sdf.Layer.FindOrOpen(layers[1])

    stage.MuteLayer(layers[0])
    stage.MuteLayer(layers[1])
    stage.UnmuteLayer(layers[1])
    stage.MuteAndUnmuteLayers([layers[2], layers[1]], [])

    # Ensure that we received the same number of notices.
    assert len(received_broker) == excepted
    assert len(received_usd) == excepted

@pytest.mark.parametrize("notice_type, expected_usd, expected_broker", [
    ("StageNotice", 12, 3),
    ("StageContentsChanged", 4, 1),
    ("ObjectsChanged", 4, 1),
    ("StageEditTargetChanged", 0, 0),
    ("LayerMutingChanged", 4, 1),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_mute_layers_batching(
    notice_type, expected_usd, expected_broker, stage_with_layers
):
    """Mute several layers and batch broker notices.
    """
    stage = stage_with_layers
    broker = Broker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(UnfNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    broker.BeginTransaction()

    layers = stage.GetRootLayer().subLayerPaths

    # Keep ref pointer to the layer we try to mute and unmute to
    # prevent it for being destroyed when it is muted.
    _layer = Sdf.Layer.FindOrOpen(layers[1])

    stage.MuteLayer(layers[0])
    stage.MuteLayer(layers[1])
    stage.UnmuteLayer(layers[1])
    stage.MuteAndUnmuteLayers([layers[2], layers[1]], [])

    # Ensure that broker notices are not sent during a transaction.
    assert len(received_broker) == 0

    # While USD Notices are being sent as expected.
    assert len(received_usd) == expected_usd

    broker.EndTransaction()

    # Ensure that consolidated broker notices are sent after a transaction.
    assert len(received_broker) == expected_broker

@pytest.mark.parametrize("notice_type, expected_usd", [
    ("StageNotice", 12),
    ("StageContentsChanged", 4),
    ("ObjectsChanged", 4),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 4),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_mute_layers_blocking(
    notice_type, expected_usd, stage_with_layers
):
    """Mute several layers and block broker notices.
    """
    stage = stage_with_layers
    broker = Broker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(UnfNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    # Predicate blocking all broker notices.
    broker.BeginTransaction(predicate=lambda _: False)

    layers = stage.GetRootLayer().subLayerPaths

    # Keep ref pointer to the layer we try to mute and unmute to
    # prevent it for being destroyed when it is muted.
    _layer = Sdf.Layer.FindOrOpen(layers[1])

    stage.MuteLayer(layers[0])
    stage.MuteLayer(layers[1])
    stage.UnmuteLayer(layers[1])
    stage.MuteAndUnmuteLayers([layers[2], layers[1]], [])

    # Ensure that broker notices are not sent during a transaction.
    assert len(received_broker) == 0

    # While USD Notices are being sent as expected.
    assert len(received_usd) == expected_usd

    broker.EndTransaction()

    # Ensure that no broker notices have been received.
    assert len(received_broker) == 0

def test_mute_layers_transaction_objectschanged(stage_with_layers):
    """Mute several layers during transaction and analyze ObjectsChanged notice.
    """
    stage = stage_with_layers
    broker = Broker.Create(stage)

    received = []

    def _validate(notice, stage):
        """Validate notice received."""
        assert len(notice.GetResyncedPaths()) == 1
        assert len(notice.GetChangedInfoOnlyPaths()) == 0
        assert "/" in notice.GetResyncedPaths()
        received.append(notice)

    # Create prim before caching to trigger resync path when muting.
    layers = stage.GetRootLayer().subLayerPaths
    layer = Sdf.Layer.Find(layers[0])
    stage.SetEditTarget(Usd.EditTarget(layer))
    stage.DefinePrim("/Foo")

    layer2 = Sdf.Layer.Find(layers[1])
    stage.SetEditTarget(Usd.EditTarget(layer2))
    stage.DefinePrim("/Bar")

    key = Tf.Notice.Register(UnfNotice.ObjectsChanged, _validate, stage)

    broker.BeginTransaction()

    stage.MuteLayer(layers[0])
    stage.MuteLayer(layers[1])
    stage.UnmuteLayer(layers[1])
    stage.MuteAndUnmuteLayers([layers[2], layers[1]], [])

    broker.EndTransaction()

    # Ensure that one notice was received.
    assert len(received) == 1

def test_mute_layers_transaction_layermutingchanged(stage_with_layers):
    """Mute several layers during transaction and analyze
    LayerMutingChanged notice.

    """
    stage = stage_with_layers
    layers = stage.GetRootLayer().subLayerPaths

    broker = Broker.Create(stage)

    received = []

    def _validate(notice, stage):
        """Validate notice received."""
        assert len(notice.GetMutedLayers()) == 3
        assert len(notice.GetUnmutedLayers()) == 0
        assert notice.GetMutedLayers()[0] == layers[0]
        assert notice.GetMutedLayers()[1] == layers[2]
        assert notice.GetMutedLayers()[2] == layers[1]
        received.append(notice)

    key = Tf.Notice.Register(UnfNotice.LayerMutingChanged, _validate, stage)

    broker.BeginTransaction()

    # Keep ref pointer to the layer we try to mute and unmute to
    # prevent it for being destroyed when it is muted.
    _layer = Sdf.Layer.FindOrOpen(layers[1])

    stage.MuteLayer(layers[0])
    stage.MuteLayer(layers[1])
    stage.UnmuteLayer(layers[1])
    stage.MuteAndUnmuteLayers([layers[2], layers[1]], [])

    broker.EndTransaction()

    # Ensure that one notice was received.
    assert len(received) == 1

File: test/integration/testMuteLayers.cpp
#include <unf/broker.h>

#include <unfTest/listener.h>
#include <unfTest/notice.h>
#include <unfTest/observer.h>

#include <gtest/gtest.h>
#include <pxr/usd/sdf/layer.h>
#include <pxr/usd/sdf/path.h>
#include <pxr/usd/usd/primRange.h>
#include <pxr/usd/usd/stage.h>

#include <string>
#include <typeinfo>

// namespace aliases for convenience.
using _USD = PXR_NS::UsdNotice;
namespace _Broker = unf::BrokerNotice;

class MuteLayersTest : public ::testing::Test {
  protected:
    using UsdListener = ::Test::Listener<
        _USD::StageNotice, _USD::StageContentsChanged, _USD::ObjectsChanged,
        _USD::StageEditTargetChanged, _USD::LayerMutingChanged>;

    using BrokerListener = ::Test::Listener<
        _Broker::StageNotice, _Broker::StageContentsChanged,
        _Broker::ObjectsChanged, _Broker::StageEditTargetChanged,
        _Broker::LayerMutingChanged>;

    void SetUp() override
    {
        _stage = PXR_NS::UsdStage::CreateInMemory();
        _AddLayers(3);

        _usdListener.SetStage(_stage);
        _brokerListener.SetStage(_stage);
    }

    void _AddLayers(int number)
    {
        _layers.reserve(number);
        _layerIds.reserve(number);

        while (_layers.size() < number) {
            auto layer = PXR_NS::SdfLayer::CreateAnonymous(".usda");
            _layers.push_back(layer);
            _layerIds.push_back(layer->GetIdentifier());
        }

        _stage->GetRootLayer()->SetSubLayerPaths(_layerIds);
    }

    PXR_NS::UsdStageRefPtr _stage;
    std::vector<PXR_NS::SdfLayerRefPtr> _layers;
    std::vector<std::string> _layerIds;

    UsdListener _usdListener;
    BrokerListener _brokerListener;
};

TEST_F(MuteLayersTest, Simple)
{
    auto broker = unf::Broker::Create(_stage);

    _stage->MuteLayer(_layerIds[0]);
    _stage->MuteLayer(_layerIds[1]);
    _stage->UnmuteLayer(_layerIds[1]);
    _stage->MuteAndUnmuteLayers(
        std::vector<std::string>{_layerIds[2], _layerIds[1]},
        std::vector<std::string>{});

    // Ensure that similar notices are received via the stage and the broker.
    ASSERT_EQ(_usdListener.Received<_USD::StageNotice>(), 12);
    ASSERT_EQ(_usdListener.Received<_USD::StageContentsChanged>(), 4);
    ASSERT_EQ(_usdListener.Received<_USD::ObjectsChanged>(), 4);
    ASSERT_EQ(_usdListener.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_usdListener.Received<_USD::LayerMutingChanged>(), 4);

    ASSERT_EQ(_brokerListener.Received<_Broker::StageNotice>(), 12);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageContentsChanged>(), 4);
    ASSERT_EQ(_brokerListener.Received<_Broker::ObjectsChanged>(), 4);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::LayerMutingChanged>(), 4);
}

TEST_F(MuteLayersTest, Batching)
{
    auto broker = unf::Broker::Create(_stage);

    broker->BeginTransaction();

    _stage->MuteLayer(_layerIds[0]);
    _stage->MuteLayer(_layerIds[1]);
    _stage->UnmuteLayer(_layerIds[1]);
    _stage->MuteAndUnmuteLayers(
        std::vector<std::string>{_layerIds[2], _layerIds[1]},
        std::vector<std::string>{});

    // Ensure that USD Notices are being sent as expected.
    ASSERT_EQ(_usdListener.Received<_USD::StageNotice>(), 12);
    ASSERT_EQ(_usdListener.Received<_USD::StageContentsChanged>(), 4);
    ASSERT_EQ(_usdListener.Received<_USD::ObjectsChanged>(), 4);
    ASSERT_EQ(_usdListener.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_usdListener.Received<_USD::LayerMutingChanged>(), 4);

    // While broker notices are blocked during a transaction.
    ASSERT_EQ(_brokerListener.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::LayerMutingChanged>(), 0);

    broker->EndTransaction();

    // Ensure that consolidated broker notices are sent after a transaction.
    ASSERT_EQ(_brokerListener.Received<_Broker::StageNotice>(), 3);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageContentsChanged>(), 1);
    ASSERT_EQ(_brokerListener.Received<_Broker::ObjectsChanged>(), 1);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::LayerMutingChanged>(), 1);
}

TEST_F(MuteLayersTest, Blocking)
{
    auto broker = unf::Broker::Create(_stage);

    // Pass a predicate to block all broker notices.
    broker->BeginTransaction([](const _Broker::StageNotice&) { return false; });

    _stage->MuteLayer(_layerIds[0]);
    _stage->MuteLayer(_layerIds[1]);
    _stage->UnmuteLayer(_layerIds[1]);
    _stage->MuteAndUnmuteLayers(
        std::vector<std::string>{_layerIds[2], _layerIds[1]},
        std::vector<std::string>{});

    // Ensure that USD Notices are being sent as expected.
    ASSERT_EQ(_usdListener.Received<_USD::StageNotice>(), 12);
    ASSERT_EQ(_usdListener.Received<_USD::StageContentsChanged>(), 4);
    ASSERT_EQ(_usdListener.Received<_USD::ObjectsChanged>(), 4);
    ASSERT_EQ(_usdListener.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_usdListener.Received<_USD::LayerMutingChanged>(), 4);

    // While broker notices are blocked during a transaction.
    ASSERT_EQ(_brokerListener.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::LayerMutingChanged>(), 0);

    broker->EndTransaction();

    // Ensure that no broker notices are sent after a transaction either.
    ASSERT_EQ(_brokerListener.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::LayerMutingChanged>(), 0);
}

TEST_F(MuteLayersTest, PartialBlocking)
{
    auto broker = unf::Broker::Create(_stage);

    std::string target = typeid(_Broker::LayerMutingChanged).name();

    // Pass a predicate to block all broker notices.
    broker->BeginTransaction([&](const _Broker::StageNotice& n) {
        return (n.GetTypeId() == target);
    });

    _stage->MuteLayer(_layerIds[0]);
    _stage->MuteLayer(_layerIds[1]);
    _stage->UnmuteLayer(_layerIds[1]);
    _stage->MuteAndUnmuteLayers(
        std::vector<std::string>{_layerIds[2], _layerIds[1]},
        std::vector<std::string>{});

    // Ensure that USD Notices are being sent as expected.
    ASSERT_EQ(_usdListener.Received<_USD::StageNotice>(), 12);
    ASSERT_EQ(_usdListener.Received<_USD::StageContentsChanged>(), 4);
    ASSERT_EQ(_usdListener.Received<_USD::ObjectsChanged>(), 4);
    ASSERT_EQ(_usdListener.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_usdListener.Received<_USD::LayerMutingChanged>(), 4);

    // While broker notices are blocked during a transaction.
    ASSERT_EQ(_brokerListener.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::LayerMutingChanged>(), 0);

    broker->EndTransaction();

    // Ensure that only consolidated LayerMutingChanged broker notice are sent.
    ASSERT_EQ(_brokerListener.Received<_Broker::StageNotice>(), 1);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_Broker::LayerMutingChanged>(), 1);
}

TEST_F(MuteLayersTest, Transaction_ObjectsChanged)
{
    auto broker = unf::Broker::Create(_stage);

    // Create prim before caching to trigger resync path when muting.
    _stage->SetEditTarget(PXR_NS::UsdEditTarget(_layers[0]));
    _stage->DefinePrim(PXR_NS::SdfPath{"/Foo"});

    ::Test::Observer<_Broker::ObjectsChanged> observer(_stage);

    broker->BeginTransaction();

    _stage->MuteLayer(_layerIds[0]);
    _stage->MuteLayer(_layerIds[1]);
    _stage->UnmuteLayer(_layerIds[1]);
    _stage->MuteAndUnmuteLayers(
        std::vector<std::string>{_layerIds[2], _layerIds[1]},
        std::vector<std::string>{});

    ASSERT_EQ(observer.Received(), 0);

    broker->EndTransaction();

    ASSERT_EQ(observer.Received(), 1);

    const auto& n = observer.GetLatestNotice();
    ASSERT_EQ(n.GetResyncedPaths().size(), 1);
    ASSERT_EQ(n.GetResyncedPaths().at(0), PXR_NS::SdfPath{"/"});
    ASSERT_EQ(n.GetChangedInfoOnlyPaths().size(), 0);
}

TEST_F(MuteLayersTest, Transaction_LayerMutingChanged)
{
    auto broker = unf::Broker::Create(_stage);

    ::Test::Observer<_Broker::LayerMutingChanged> observer(_stage);

    broker->BeginTransaction();

    _stage->MuteLayer(_layerIds[0]);
    _stage->MuteLayer(_layerIds[1]);
    _stage->UnmuteLayer(_layerIds[1]);
    _stage->MuteAndUnmuteLayers(
        std::vector<std::string>{_layerIds[2], _layerIds[1]},
        std::vector<std::string>{});

    ASSERT_EQ(observer.Received(), 0);

    broker->EndTransaction();

    ASSERT_EQ(observer.Received(), 1);

    const auto& n = observer.GetLatestNotice();
    auto layerIds = _stage->GetRootLayer()->GetSubLayerPaths();
    ASSERT_EQ(n.GetMutedLayers().size(), 3);
    ASSERT_EQ(n.GetMutedLayers().at(0), std::string(layerIds[0]));
    ASSERT_EQ(n.GetMutedLayers().at(1), std::string(layerIds[2]));
    ASSERT_EQ(n.GetMutedLayers().at(2), std::string(layerIds[1]));
    ASSERT_EQ(n.GetUnmutedLayers().size(), 0);
}
